src/viam/ur/module/ur_arm.cpp

```diff
@@ -172,128 +184,1284 @@ void write_waypoints_to_csv(const std::string& filepath, const std::list<Eigen::
+}
+
+auto URArm::state_::event_remote_mode_restored_::describe() const {
+    return name();
+}
+
+// NOLINTEND(readability-convert-member-functions-to-static)
+
+URArm::state_::move_request::move_request(std::vector<trajectory_sample_point>&& samples, std::ofstream arm_joint_positions_stream)
+    : samples(std::move(samples)), arm_joint_positions_stream(std::move(arm_joint_positions_stream)) {
+    if (this->samples.empty()) {
+        throw std::invalid_argument("no trajectory samples provided to move request");
+    }
+}
+
+std::shared_future<void> URArm::state_::move_request::cancel() {
+    if (!cancellation_request) {
+        auto& cr = cancellation_request.emplace();
+        cr.future = cr.promise.get_future().share();
+    }
+    return cancellation_request->future;
+}
 
-    // specified through APPDIR environment variable
-    std::string appdir;
+void URArm::state_::move_request::complete_success() {
+    // Mark the move_request as completed. If there
+    // was a cancel request, it raced and lost, but it doesn't
+    // need an error.
+    completion.set_value();
+    if (cancellation_request) {
+        cancellation_request->promise.set_value();
+    }
+}
+
+void URArm::state_::move_request::complete_cancelled() {
+    complete_error("arm's current trajectory cancelled");
+}
+
+void URArm::state_::move_request::complete_failure() {
+    complete_error("arm's current trajectory failed");
+}
+
+void URArm::state_::move_request::complete_error(std::string_view message) {
+    // The trajectory is being completed with an error of some sort. Set the completion result to an error,
+    // and unblock any cancellation request.
+    completion.set_exception(std::make_exception_ptr(std::runtime_error{std::string{message}}));
+    if (cancellation_request) {
+        cancellation_request->promise.set_value();
+    }
+    VIAM_SDK_LOG(warn) << "A trajectory completed with an error: " << message;
+}
+
+void URArm::state_::move_request::cancel_error(std::string_view message) {
+    std::exchange(cancellation_request, {})->promise.set_exception(std::make_exception_ptr(std::runtime_error{std::string{message}}));
+}
+
+void URArm::state_::move_request::write_joint_data(vector6d_t& position, vector6d_t& velocity) {
+    ::write_joint_data(position, velocity, arm_joint_positions_stream, unix_time_iso8601(), arm_joint_positions_sample++);
+}
+
+URArm::state_::move_request::cancellation_request::cancellation_request() {}
+
+void URArm::state_::emit_event_(event_variant_&& event) {
+    std::visit([this](auto&& event) { this->emit_event_(std::forward<decltype(event)>(event)); }, std::move(event));
+}
 
-    // variables specified by ResourceConfig and set through reconfigure
-    std::string host;
-    std::atomic<double> speed{0};
-    std::atomic<double> acceleration{0};
-    std::atomic<bool> estop{false};
-    std::atomic<bool> local_disconnect{false};
+template <typename T>
+void URArm::state_::emit_event_(T&& event) {
```

Comment:
The core of the state machine. Visits the current state, obtains its description and a description of the event it is about to handle. It then delegates to the current state's event handler for the current event, which will return the new state that we are going to become. We then get the description of that state, log the description, and return the new state. That newly returned state replaces our current state.